File: study/sort_algorithm/bubble_sort.py
'''
冒泡排序

基本思路：

    1. 有长度为n的数组array，比较array[0]和array[1]，如果array[0]>array[1]则交换array[0]和array[1];
    2. 继续往后遍历array[i],array[i+1]，如果array[i]>array[i+1]则交换array[i]和array[i+1];
    3. 如果交换之后array[i-1]>array[i]，则继续往前交换；
    4. 直到array[k-1]<array[k]为止。

'''

def bubble_sort(lst):

    for i in range(len(lst)-1):
        for j in range(i+1,0,-1):
            if lst[j]<lst[j-1]:
                lst[j],lst[j-1] = lst[j-1],lst[j]
                #j -= 1
    print(lst)

if __name__ == "__main__":
    lst = [1,2,42,1,3,4,6,2,5,7,784,23]
    bubble_sort(lst)

File: study/sort_algorithm/bubble_sort.cpp
#include <iostream>
using namespace std;
template <typename T>

void bubble_sort(T lst[],int len)
{
	int i,j;
	T temp;
	for (i=0;i<len-1;i++)
	{
		for (j=0;j<len-1;j++)
		{
			if (lst[j]>lst[j+1])
			{
				swap(lst[j],lst[j+1]);
			}
		}
	}
}

template <typename T>//在每个使用到T类型的函数前都需要加template 
void swap(T x,T y)
{
	T temp;
	temp = x;
	x = y;
	y = temp;
}

int main()
{
	float lst[] = {11,7,6,54,4,3,32,22,1,2,3,4,5,6,7};
	int len = (int) sizeof(lst)/sizeof(*lst);
	bubble_sort(lst,len);
	for (int i=0;i<len;i++)
	{
		cout << lst[i] << " ";
	}
	cout << endl;
	return 0;
}
